src/nifreeze/model/base.py

```diff
@@ -84,111 +83,78 @@ class BaseModel:
 
     """
 
-    __slots__ = (
-        "_model",
-        "_mask",
-        "_models",
-        "_datashape",
-        "_is_fitted",
-        "_modelargs",
-    )
+    __slots__ = {
+        "_dataset": "Reference to a :obj:`~nifreeze.data.base.BaseDataset` object.",
+    }
 
-    def __init__(self, mask=None, **kwargs):
+    def __init__(self, dataset, **kwargs):
         """Base initialization."""
 
-        # Keep model state
-        self._model = None  # "Main" model
-        self._models = None  # For parallel (chunked) execution
-
-        # Setup brain mask
-        if mask is None:
+        self._dataset = dataset
+        # Warn if mask not present
+        if dataset.brainmask is None:
             warn(mask_absence_warn_msg, stacklevel=2)
 
-        self._mask = mask
-
-        self._datashape = None
-        self._is_fitted = False
-
-        self._modelargs = ()
-
-    @property
-    def is_fitted(self):
-        return self._is_fitted
-
-    def fit(self, data, **kwargs):
-        """Abstract member signature of fit()."""
-        raise NotImplementedError("Cannot call fit() on a BaseModel instance.")
-
-    def predict(self, *args, **kwargs):
-        """Abstract member signature of predict()."""
-        raise NotImplementedError("Cannot call predict() on a BaseModel instance.")
+    def fit_predict(self, *_, **kwargs):
```

Comment:
You're specifically permitting arbitrary positional args that will be thrown away? This feels like a strange API. You can disable them entirely with:

```suggestion
    def fit_predict(self, *, **kwargs):
```

That feels like less of a footgun.

RE the NotImplementedError, is this a method that will always be implemented by subclasses? If so, then marking it an `@abstractmethod` might be helpful. Given that you have a `TrivialModel` below, there's no loss to not being able to instantiate a `BaseModel()`.

If it's not always going to exist, you could remove this here, and create a Protocol that can be used to match only `BaseModel`s that implement the method:

```python
class FitPredictModel(ty.Protocol):
    def fit_predict(self, **kwargs): ...
```

Also, the current types are `Any`. Do you know enough to say something like:

```suggestion
    def fit_predict(self, *, **kwargs) -> np.ndarray:
```